sei-tendermint/internal/p2p/transport.go

```diff
@@ -80,38 +92,113 @@ func HandshakeOrClose(
 		}
 		ok.Store(true)
 		return &Connection{
-			conn:     tcpConn,
-			peerInfo: peerInfo,
+			conn:         tcpConn,
+			sendQueue:    NewQueue[sendMsg](queueBufferDefault),
+			peerInfo:     peerInfo,
+			peerChannels: toChannelIDs(peerInfo.Channels),
 			mconn: conn.NewMConnection(
-				logger.With("peer", remoteEndpoint(tcpConn).NodeAddress(peerInfo.NodeID)),
+				r.logger.With("peer", remoteEndpoint(tcpConn).NodeAddress(peerInfo.NodeID)),
 				secretConn,
-				channelDescs,
-				mConnConfig,
+				r.getChannelDescs(),
+				r.options.Connection,
 			),
 		}, nil
 	})
 }
 
-func (c *Connection) Run(ctx context.Context) error {
-	return c.mconn.Run(ctx)
+func (c *Connection) sendRoutine(ctx context.Context, r *Router) error {
+	for {
+		start := time.Now().UTC()
+		m, err := c.sendQueue.Recv(ctx)
+		if err != nil {
+			return err
+		}
+		r.metrics.RouterPeerQueueRecv.Observe(time.Since(start).Seconds())
+		bz, err := proto.Marshal(m.Message)
+		if err != nil {
+			panic(fmt.Sprintf("proto.Marshal(): %v", err))
+		}
+		if m.ChannelID > math.MaxUint8 {
+			return fmt.Errorf("MConnection only supports 1-byte channel IDs (got %v)", m.ChannelID)
+		}
+		if err := c.mconn.Send(ctx, m.ChannelID, bz); err != nil {
+			return err
+		}
+		r.logger.Debug("sent message", "peer", c.peerInfo.NodeID, "message", m.Message)
+	}
 }
 
-// String displays connection information.
-func (c *Connection) String() string {
-	return c.RemoteEndpoint().String()
+// receivePeer receives inbound messages from a peer, deserializes them and
+// passes them on to the appropriate channel.
+func (c *Connection) recvRoutine(ctx context.Context, r *Router) error {
+	for {
+		chID, bz, err := c.mconn.Recv(ctx)
+		if err != nil {
+			return err
+		}
+		for chs := range r.channels.RLock() {
+			ch, ok := chs[chID]
+			if !ok {
+				// TODO(gprusak): verify if this is a misbehavior, and drop the peer if it is.
+				r.logger.Debug("dropping message for unknown channel", "peer", c.peerInfo.NodeID, "channel", chID)
+				continue
+			}
```

Comment:
Why do we need to hold the RLock after this point if we've already read the channel itself? is this to enforce synchronization on the underlying channel as well? if so, wouldn't it make more sense to have RWMutex[*channel] as the map value and acquire that instead of locking the full map until we can send to the recvQueue?

Reply:
btw, nice catch with the `r.channels.Lock()` in getChannelDescs() above. That would be problematic indeed.